### PR TITLE
core(driver): extract waitFor methods

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Paves the way for adding the extra waitFor methods needed for #11180 in a way that still helps out with fraggle rock and makes future testing easier.

This is a move and slight restructuring to convert `this` references to parameters *only*. No functionality changes, and no substantive test additions (though the gaps are much more obvious and should be much easier to test now).

**Related Issues/PRs**
ref #11313 , #11180 , #10810

### CODE REVIEW

File: lighthouse-core/gather/driver/wait-for-condition.js [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+ */
+function waitForNothing() {
+  return {promise: Promise.resolve(), cancel() {}};
+}
+
+/**
+ * Returns a promise that resolve when a frame has been navigated.
+ * Used for detecting that our about:blank reset has been completed.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @return {CancellableWait}
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/wait-for-condition.js:41]: this is the only return type change, being more consistent felt like it belonged here even if we don't use `cancel` yet

File: lighthouse-core/gather/driver/wait-for-condition.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+/**
+ * Returns a promise that resolves when the network has been idle (after DCL) for
+ * `networkQuietThresholdMs` ms and a method to cancel internal network listeners/timeout.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {NetworkRecorder} networkMonitor
+ * @param {number} networkQuietThresholdMs
+ * @return {CancellableWait}
+ */
+function waitForNetworkIdle(session, networkMonitor, networkQuietThresholdMs) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/wait-for-condition.js:118]: extracting `session` and `networkMonitor` to parameters are the only substantive changes in this PR

File: lighthouse-core/gather/driver/wait-for-condition.js [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    };
+  }
+
+  /** @type {NodeJS.Timer|undefined} */
+  let lastTimeout;
+  let canceled = false;
+
+  const checkForQuietExpression = `(${pageFunctions.checkTimeSinceLastLongTaskString})()`;
+  /**
+   * @param {ExecutionContext} executionContext
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/wait-for-condition.js:215]: directly using `executionContext` is an FR-compat change and functionally equivalent to using driver when isolation is disabled

File: lighthouse-core/gather/driver/wait-for-condition.js [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+ * See https://github.com/GoogleChrome/lighthouse/issues/627 for more.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {NetworkRecorder} networkMonitor
+ * @param {WaitOptions} options
+ * @return {Promise<{timedOut: boolean}>}
+ */
+async function waitForFullyLoaded(session, networkMonitor, options) {
+  const {pauseAfterFcpMs, pauseAfterLoadMs, networkQuietThresholdMs,
+    cpuQuietThresholdMs, maxWaitForLoadedMs, maxWaitForFcpMs} = options;
+  const {waitForFcp, waitForLoadEvent, waitForNetworkIdle, waitForCPUIdle} =
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/wait-for-condition.js:336]: the last remaining change that exposes the underlying waits for easier testing

File: lighthouse-core/gather/driver/wait-for-condition.js [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+  resolveOnLoadEvent.cancel();
+  resolveOnNetworkIdle.cancel();
+  resolveOnCPUIdle.cancel();
+
+  return cleanupFn();
+}
+
+
+module.exports = {
+  waitForNothing,
```
Comment (connorjclark) [inline lighthouse-core/gather/driver/wait-for-condition.js:413]:
```suggestion
}

module.exports = {
  waitForNothing,
```

**Review by connorjclark — APPROVED** [review]
LGTM

**Pull request merged**